File: pca.py
import numpy as np

class PCA:
    def __init__(self, n_components):
        self.n_components = n_components
        self.mean = None
        self.eigenvectors = None

    def fit(self, data):
        self.mean = np.mean(data, axis = 0)
        normalized_data = data - self.mean
        covariance_matrix = np.cov(normalized_data, rowvar = False)
        eigenvalues, self.eigenvectors = np.linalg.eig(covariance_matrix)
        sorted_indices = np.argsort(eigenvalues)[::-1]
        self.eigenvectors = self.eigenvectors[:, sorted_indices]

    def transform(self, data):
        normalized_data = data - self.mean
        reduced_data = np.dot(normalized_data, self.eigenvectors[:, :self.n_components])
        return reduced_data

    def inverse_transform(self, reduced_data):
        reconstructed_data = np.dot(reduced_data, self.eigenvectors[:, :self.n_components].T) + self.mean
        return reconstructed_data


data = np.random.random(size=(100,3))

pca = PCA(2)

pca.fit(data)

pca.transform(data)

File: pca.jl
using Statistics
using LinearAlgebra

struct PCA

    n_components :: Int
    mean :: Vector{Float64}
    eigenvectors :: Matrix{Float64}

    function PCA(data, n_components)

        μ = vec(mean(data, dims = 1))
        normalized_data = data .- μ'
        covariance_matrix = cov(normalized_data)
        eigenvalues, eigenvectors = eigen(covariance_matrix)
        sorted_indices = sortperm(eigenvalues, rev = true)

        new( n_components, μ, eigenvectors[:, sorted_indices[1:n_components]])

    end

end

function transform(pca, data)
    normalized_data = data .- pca.mean'
    reduced_data = normalized_data * pca.eigenvectors
    return reduced_data
end

function inverse_transform!(pca, reduced_data)
    reconstructed_data = reduced_data * pca.eigenvectors
    return reconstructed_data
end


data = rand(100,3)

pca = PCA(data, 2)

transform(pca, data)


